File: {{cookiecutter.project_slug}}/src/{{cookiecutter.project_slug}}/auth/oauth.py
{% if cookiecutter.sdk_choice == 'python' and cookiecutter.auth_mechanism == 'oauth2' -%}
"""OAuth 2.1 authentication handler."""

import logging
from typing import Optional
from authlib.integrations.starlette_client import OAuth
from starlette.requests import Request

logger = logging.getLogger(__name__)


class OAuthHandler:
    """OAuth 2.1 authentication handler following MCP best practices."""

    def __init__(
        self,
        client_id: Optional[str] = None,
        client_secret: Optional[str] = None,
        authorization_endpoint: Optional[str] = None,
        token_endpoint: Optional[str] = None,
    ):
        """Initialize OAuth handler.

        Args:
            client_id: OAuth client ID
            client_secret: OAuth client secret (use PKCE for public clients)
            authorization_endpoint: OAuth authorization URL
            token_endpoint: OAuth token URL
        """
        self.client_id = client_id
        self.client_secret = client_secret
        self.authorization_endpoint = authorization_endpoint
        self.token_endpoint = token_endpoint

        # Initialize OAuth client
        self.oauth = OAuth()
        if client_id:
            self.oauth.register(
                name='mcp_provider',
                client_id=client_id,
                client_secret=client_secret,
                authorize_url=authorization_endpoint,
                access_token_url=token_endpoint,
                client_kwargs={'scope': 'openid profile email'},
            )

    async def authenticate(self, request: Request) -> bool:
        """Authenticate incoming request.

        Args:
            request: Starlette request object

        Returns:
            True if authenticated, False otherwise
        """
        # Extract bearer token from Authorization header
        auth_header = request.headers.get('Authorization', '')
        if not auth_header.startswith('Bearer '):
            logger.warning("Missing or invalid Authorization header")
            return False

        token = auth_header[7:]  # Remove 'Bearer ' prefix

        try:
            # Verify token (implement your token verification logic)
            # This could involve introspection endpoint or JWT validation
            await self._verify_token(token)
            return True
        except Exception as e:
            logger.error(f"Token verification failed: {e}")
            return False

    async def _verify_token(self, token: str) -> None:
        """Verify OAuth token.

        Args:
            token: Bearer token to verify

        Raises:
            ValueError: If token is invalid
        """
        # Implement token verification logic here
        # Options:
        # 1. Call OAuth provider's introspection endpoint
        # 2. Validate JWT signature if using JWT tokens
        # 3. Check token against local cache/database
        pass

    def get_authorization_url(self) -> str:
        """Get OAuth authorization URL for user consent.

        Returns:
            Authorization URL for user to grant consent
        """
        if not self.authorization_endpoint:
            raise ValueError("Authorization endpoint not configured")

        # Build authorization URL with PKCE for public clients
        # This should be called from your UI to redirect users
        return f"{self.authorization_endpoint}?client_id={self.client_id}&response_type=code&scope=openid"
{%- endif %}

File: {{cookiecutter.project_slug}}/src/auth/oauth.ts
{% if cookiecutter.sdk_choice == 'typescript' and cookiecutter.auth_mechanism == 'oauth2' -%}
/**
 * OAuth 2.1 authentication handler
 * Follows MCP security best practices
 */

import { Request } from "express";
import { Issuer, Client } from "openid-client";

const log = {
  info: (msg: string) => console.error(`[INFO] ${msg}`),
  error: (msg: string) => console.error(`[ERROR] ${msg}`),
  warn: (msg: string) => console.error(`[WARN] ${msg}`),
};

export class OAuthHandler {
  private client?: Client;

  constructor(
    private clientId?: string,
    private clientSecret?: string,
    private issuerUrl?: string
  ) {
    this.initialize();
  }

  private async initialize(): Promise<void> {
    if (!this.issuerUrl || !this.clientId) {
      log.warn("OAuth not configured - authentication will fail");
      return;
    }

    try {
      const issuer = await Issuer.discover(this.issuerUrl);
      this.client = new issuer.Client({
        client_id: this.clientId,
        client_secret: this.clientSecret,
        token_endpoint_auth_method: this.clientSecret ? "client_secret_basic" : "none",
      });
      log.info("OAuth client initialized");
    } catch (error) {
      log.error(`Failed to initialize OAuth: ${error}`);
    }
  }

  async authenticate(req: Request): Promise<boolean> {
    const authHeader = req.headers.authorization;

    if (!authHeader || !authHeader.startsWith("Bearer ")) {
      log.warn("Missing or invalid Authorization header");
      return false;
    }

    const token = authHeader.substring(7);

    try {
      await this.verifyToken(token);
      return true;
    } catch (error) {
      log.error(`Token verification failed: ${error}`);
      return false;
    }
  }

  private async verifyToken(token: string): Promise<void> {
    if (!this.client) {
      throw new Error("OAuth client not initialized");
    }

    // Verify token via introspection endpoint or JWT validation
    // Implementation depends on your OAuth provider
    const userinfo = await this.client.userinfo(token);
    log.info(`Token verified for user: ${userinfo.sub}`);
  }

  getAuthorizationUrl(): string {
    if (!this.client) {
      throw new Error("OAuth client not initialized");
    }

    // Generate authorization URL with PKCE for public clients
    const url = this.client.authorizationUrl({
      scope: "openid profile email",
      response_type: "code",
    });

    return url;
  }
}
{%- endif %}
